lib/puppet/settings.rb

```diff
@@ -385,6 +385,19 @@ def initialize_app_defaults(app_defaults)
     call_hooks_deferred_to_application_initialization
     issue_deprecations
 
+    run_mode = Puppet::Util::RunMode[self.preferred_run_mode]
+    if run_mode.agent? || run_mode.server?
+      if self.set_in_section?(:masterport, run_mode.name) && !self.set_in_section?(:serverport, run_mode.name)
+        self[:serverport] = self[:masterport]
+      elsif self.set_by_config?(:masterport) && !self.set_by_config?(:serverport)
+        self[:serverport] = self[:masterport]
```

Comment:
notes to future self:

if `:masterport` is set in `main`, then its hook will be called to set `serverport` to the same value, and `set_by_config?(:serverport)` will return true. 

If `:masterport` is set in `agent`, then its hook won't be called (because the hook type is `:on_write_only`), and `set_by_config?(:serverport)` will return false, in which case we want to set `:serverport` here.

Reply:
> If `:masterport` is set in `agent`, then its hook won't be called (because the hook type is `:on_write_only`), and `set_by_config?(:serverport)` will return false, in which case we want to set `:serverport` here.

More on the problem of hooks not being called:

Settings with hooks are collected and evaluated when the config file is parsed (before the application is initialized):
https://github.com/puppetlabs/puppet/blob/03073d2285ce32011b7a401e04b83266c12d4a13/lib/puppet/settings.rb#L609

The line above gets all value sets, taking into note the current environment and run mode (which is *always* `:user` at this point, propagated all the way from here: https://github.com/puppetlabs/puppet/blob/03073d2285ce32011b7a401e04b83266c12d4a13/lib/puppet.rb#L153). The `value_sets_for` method ends up calling `configsearchpath` which constructs the search order with the following priority (`:memory`, `:cli`, environment, run mode (always `:user` here), `:main`). So if the setting is in a section different than `:main`, or `:user`, it won't be collected here, and its hook will not be called.

This also confuses `puppet config`, as it won't correctly evaluate settings with hooks in sections other than `:main` or `:user`. A way to fix this would be to ensure whatever the user passes in the `--section` param is in the config search path.

Reply:
> So if the setting is in a section different than :main, or :user, it won't be collected here, and its hook will not be called.

We also register `:on_initialize_and_write` hooks when the setting is defined https://github.com/puppetlabs/puppet/blob/e68bf5794e213597af93df558d4a16216ec94078/lib/puppet/settings.rb#L991

So that later `Puppet::Settings.initialize_app_defaults` calls all `hooks_to_call_on_application_initialization` https://github.com/puppetlabs/puppet/blob/e68bf5794e213597af93df558d4a16216ec94078/lib/puppet/settings.rb#L411

This can be seen by changing the hook for `node_name_fact` to `:on_initialize_and_write` and adding the following to the `[agent]` section in puppet.conf:

```ini
[agent]
node_name_value=a
certname=b
node_name_fact=c
```

in which case the hook will be called.

> This also confuses puppet config as it won't correctly evaluate settings with hooks in sections other than :main or :user

That's a good point. I'll make a note of that in PR #8400

Reply:
About the deferred application hooks, this test shows how the `on_write_only` hooks (default) are not called if the setting is set in `agent` section: https://github.com/puppetlabs/puppet/pull/8395/files#diff-634da12997fa34484185d76fb4caf7bfddcffa7ac6d0597ca7289590186e8939R529

And same with `:on_define_and_write`: https://github.com/puppetlabs/puppet/pull/8395/files#diff-634da12997fa34484185d76fb4caf7bfddcffa7ac6d0597ca7289590186e8939R603-R604

But they are called for `:on_initialize_and_write` hooks: https://github.com/puppetlabs/puppet/pull/8395/files#diff-634da12997fa34484185d76fb4caf7bfddcffa7ac6d0597ca7289590186e8939R674

Reply:
> About the deferred application hooks, this test shows how the `on_write_only` hooks (default) are not called if the setting is set in `agent` section: [#8395 (files)](https://github.com/puppetlabs/puppet/pull/8395/files#diff-634da12997fa34484185d76fb4caf7bfddcffa7ac6d0597ca7289590186e8939R529)

This part is a bit strange to me, as I would have expected the `on_write_only` hook to behave the same regardless of setting section. Shouldn't it?

This can be done in the `initialize_app_defaults` method, after the run mode is changed (basically redo the thing from here https://github.com/puppetlabs/puppet/blob/e68bf5794e213597af93df558d4a16216ec94078/lib/puppet/settings.rb#L608 but for specifically for our run_mode/section).

Reply:
> I would have expected the on_write_only hook to behave the same regardless of setting section

I agree that it should behave that way, but unfortunately it doesn't currently, as each setting that wants the "correct" behavior needs to specify `on_initialize_and_write`.

Could we just get rid of `on_write_only` hooks and switch the 6 or so settings over to using `on_initialize_and_write` instead? Then we'd be left with either `on_define_and_write` and `on_initialize_and_write`.

Is there ever a case where we only want a hook to execute when the global config is parsed, but not (again) after application defaults are initialized? The only one I know of is `Puppet[:path]` which is already defined as `on_define_and_write`.

Here's a topic branch with this idea: https://github.com/puppetlabs/puppet/compare/f7226f62ba...joshcooper:cleanup_hooks